### PR TITLE
Test for LeafletFinder selection string

### PR DESCRIPTION
**Pull request by tylerjereddy**
This PR adds a unit test that ensures that `LeafletFinder` returns the same result when fed a selection string instead of the atomgroup proper that is generated by that selection string. This is something a user would commonly do and matches [this missing coverage line](https://coveralls.io/builds/7924214/source?filename=miniconda%2Fenvs%2Fpyenv%2Flib%2Fpython2.7%2Fsite-packages%2FMDAnalysis%2Fanalysis%2Fleaflet.py#L117).

For clarity, we are testing to ensure that the following two statements are equivalent:

```python
leaflets = LeafletFinder(universe, 'name PO4')

leaflets = LeafletFinder(universe, universe.select_atoms('name PO4'))
```

Other notes from looking at the testing module involved:

- there's a `setUp` object that isn't being torn down after (fix this elsewhere if you want to)
- it is slightly awkward that we now have tests in the test class that both separately import `LeafletFinder`, however I believe this import mechanism is being used to skirt the scipy dependency issue for some build scenarios

### CODE REVIEW

**Comment by tylerjereddy** [comment]
Resolved merge conflicts -- this made the file diffs a bit less clean.

File: testsuite/MDAnalysisTests/test_selections.py [hunk 1]
```diff
@@ -61,6 +61,9 @@ def _write_selection(self, **kwargs):
         g.write_selection(self.namedfile, **kwargs)
         return g
 
+    def test_write_bad_mode(self):
```
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/test_selections.py]: That code has already been merged into develop, see: https://github.com/MDAnalysis/mdanalysis/pull/989

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py [hunk 2]
```diff
@@ -42,3 +44,11 @@ def test_leaflet_finder(self):
...
 
 
+    def test_string_vs_atomgroup_proper(self):
+        from MDAnalysis.analysis.leaflet import LeafletFinder
+        lfls_ag = LeafletFinder(self.universe, self.lipid_heads, pbc=True)
+        lfls_string = LeafletFinder(self.universe, self.lipid_head_string, pbc=True)
+        groups_ag = lfls_ag.groups()
+        groups_string = lfls_string.groups()
+        assert_equal(groups_string[0].indices, groups_ag[0].indices)
+        assert_equal(groups_string[1].indices, groups_ag[1].indices)
```
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: Right, another PR for that specific unit test could always be opened, and not just for this string vs. ag test, but for the general behaviour of `LeafletFinder` -- i.e., there is currently only a single unit test for this module in the code base.

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py [hunk 3]
```diff
@@ -27,9 +27,11 @@ class TestLeafletFinder(TestCase):
     def setUp(self):
         self.universe = MDAnalysis.Universe(Martini_membrane_gro, Martini_membrane_gro)
         self.lipid_heads = self.universe.select_atoms("name PO4")
+        self.lipid_head_string = "name PO4"
```
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: There is probably some merit to defining the string only once. I have a habit of moving as many variables as I can to the explicit `setUp` and `tearDown` methods, to keep the test methods proper as clean as possible, at least in cases where the initializing of the variables is not particularly time consuming.

Otherwise, I'd probably split out to a new testing class and / or use classmethods so the setup and teardown only happen once for the test class.

**Comment by tylerjereddy** [comment]
I think I did a rebase against develop and then pulled from this branch on github back to the local branch and merged that in & then pushed back to here, which is not exactly winning points for clarity, but is theoretically ok I think.

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py (hunk 2 above)
Comment (jbarnoud) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: My point was more about making sure that the group where actually the sames in the two cases (with an AtomGroup and with a selection string). A simple `assert_equal(len(groups_ag), len(group_string))` would make sure that the selection string does not spawn an unexpected leaflet.

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py [hunk 4]
```diff
@@ -27,9 +27,11 @@ class TestLeafletFinder(TestCase):
...
 
     def tearDown(self):
         del self.universe
+        del self.lipid_head_string
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: I wouldn't bother `del`ing a string. We only do this for universes and other big things to help the garbage collector or to help with closing files during test runs where objects seem to be kept alive for the whole test.

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py [hunk 5]
```diff
@@ -42,3 +44,11 @@ def test_leaflet_finder(self):
         assert_equal(bottom_heads.indices, np.arange(2521,4670,12), err_msg="Found wrong leaflet lipids")
 
 
+    def test_string_vs_atomgroup_proper(self):
+        from MDAnalysis.analysis.leaflet import LeafletFinder
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: Are we doing the in-method imports to work around some potentially missing modules? Otherwise it should be at the top of the file.

File: testsuite/MDAnalysisTests/test_selections.py (hunk 1 above)
Comment (orbeckst) [inline testsuite/MDAnalysisTests/test_selections.py]: @jbarnoud,  `a+` is a valid mode for [open](https://docs.python.org/2/library/functions.html#open) but not for these writers. The tests check both that the proper exceptions are raised and, I guess, that we haven't implemented something new. Although I see your point , I am quite fine with testing for `a+` and moving on ;-).

File: testsuite/CHANGELOG [hunk 6]
```diff
@@ -14,8 +14,12 @@ and https://github.com/MDAnalysis/mdanalysis/wiki/UnitTests
 
 ------------------------------------------------------------------------------
 ??/??/16 jbarnoud, orbeckst, fiona-naughton, manuel.nuno.melo, richardjgowers
+         tyler.je.reddy
 
   * 0.15.1
+    - added test for LeafletFinder handling a selection string
```
Comment (orbeckst) [inline testsuite/CHANGELOG]: Can you please just update 0.15.1 to 0.16.0? I forgot to do it in an earlier PR and it seems not worth a fresh commit...

**Comment by orbeckst** [comment]
Regarding rebasing etc – whatever works. I will squash everything into a single tidy commit unless there are distinct functional commits.

**Comment by tylerjereddy** [comment]
Closing this -- if the indices of particles in the first two leaflets are different but the total number of leaflets detected is the same, the latter information would be obscured by a design decision to stack multiple assertions in a single test.

Perhaps a more detailed PR with multiple conceptually-focused unit tests may be opened at another time.

File: testsuite/CHANGELOG (hunk 6 above)
Comment (tylerjereddy) [inline testsuite/CHANGELOG]: Done elsewhere: https://github.com/MDAnalysis/mdanalysis/pull/996

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py (hunk 5 above)
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: I was simply copying the in-method import strategy already in place for this unit testing module, which I assumed was justified if the code was already using that approach. The testing module does attempt to skip if `scipy` is not available, which probably explains it. I thought about placing the import statement in the `setUp` method, but then I figured there must be some justification for not having done that in the first place.

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py (hunk 4 above)
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: I like to be explicit about the unit test clean up process. Seems like a matter of personal preference.

File: testsuite/MDAnalysisTests/test_selections.py (hunk 1 above)
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/test_selections.py]: Or you could just open another PR with a more obviously-wrong value in an additional test--testing for a valid standard library open-based flag and an invalid one would serve two slightly different purposes.

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py (hunk 2 above)
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: See comment / closure note below re: stacking assertions with conceptually different tests.

**Comment by jbarnoud** [comment]
If the number of groups differs while the content of the groups remain the same is an extreme scenario. If it happens, something went terribly wrong. Therefore, I am not very worried about the test case being hidden behind the two other assertions.

I would rather merge your test than loosing it because of my tantrum about a improbable edge case.

**Comment by orbeckst** [comment]
Thanks for explaining. 

I think it's fair to say that it's always worthwhile to rethink existing code, especially if one runs into more ancient pieces. 

--
Oliver Beckstein
email: orbeckst@gmail.com

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py (hunk 5 above)
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: Thanks for explaining. I think it's fair to say that it's always worthwhile to rethink existing code, especially if one runs into more ancient pieces.

**Comment by orbeckst** [comment]
I don't quite understand why you closed the PR. The tests as they are are already valid and a useful contribution.

If you don't have the time to continue on the PR then still leave it open, at the "help" tag, and wait for someone else to continue.

Just my $0.02.

**Comment by tylerjereddy** [comment]
The discussion points border on refactoring the (apparently old?) testing module, so perhaps another PR focused on the vision you have for that would be more appropriate first! I was just trying to increase coverage by 0.007 %, not cover all the angles, so I think that's best done elsewhere with a checklist-style thing maybe.

Do feel free to merge the unit test as-is if you'd rather not lose it though, but it may just introduce technical debt based on the feedback above.

**Comment by jbarnoud** [comment]
I fixed the conflict locally. I am going to push to the branch and merge into develop. Hopefully, nothing will break...

File: package/MDAnalysis/analysis/distances.py [hunk 7]
```diff
@@ -192,6 +193,6 @@ def between(group, A, B, distance):
     from MDAnalysis.core.AtomGroup import AtomGroup
 
     ns_group = AtomNeighborSearch(group)
-    resA = set(ns_group.search_list(A, distance))
-    resB = set(ns_group.search_list(B, distance))
+    resA = set(ns_group.search(A, distance))
```
Comment (jbarnoud) [inline package/MDAnalysis/analysis/distances.py]: Why are these changes visible here? They got merged in #998! What did I break?

Comment (tylerjereddy) [inline package/MDAnalysis/analysis/distances.py]: My suspicion for the original issue was that I rebased against develop, but it wasn't actually the tip (most recent commit) of develop, and apparently that will mess up the diff that shows up here. And indeed the github diffs here are now showing all kinds of stuff that was merged recently, so it definitely looks like a rebase against an old commit level of develop.

**Comment by jbarnoud** [comment]
I'll see tomorrow what I can do about it. I also have a look at @orbeckst comment about toplevel imports.

With the new review system, asking for changes really block the merge button.

**Comment by orbeckst** [comment]
@jbarnoud just check that the merge won't create a mess – I approved the changes so far.

**Comment by tylerjereddy** [comment]
Looks like the diff is clean now after rebase and force pushing.
